haystack/nodes/ranker/lost_in_the_middle.py

```diff
@@ -0,0 +1,126 @@
+from typing import Optional, Union, List
+import logging
+
+from haystack.schema import Document
+from haystack.nodes.ranker.base import BaseRanker
+
+logger = logging.getLogger(__name__)
+
+
+class LostInTheMiddleRanker(BaseRanker):
+    """
+    The LostInTheMiddleRanker implements a ranker that reorders documents based on the lost in the middle order.
+    "Lost in the Middle: How Language Models Use Long Contexts" by Liu et al. aims to layout paragraphs into LLM
+    context so that relevant paragraphs are at the beginning or end of the input context, while the least relevant
+    information should be in the middle of a context.
+
+    See https://arxiv.org/abs/2307.03172 for more details.
+    """
+
+    def __init__(self, word_count_threshold: Optional[int] = None, truncate_document: Optional[bool] = False):
+        """
+        Creates an instance of LostInTheMiddleRanker.
+
+        If truncate_document is set to True, you must specify a word_count_threshold as well.
+
+        :param word_count_threshold: The maximum number of words in all ordered documents.
```

Comment:
Do we really want the full first document without truncation even if it exceeds the `word_count_threshold`? If so, then I understand why we need both parameters.

Reply:
Yeah, I'm not sure myself. My guideline was a seeming principle we follow - rankers shouldn't change documents. So you really have to set `truncate_document` to True to truncate it. Otherwise, we'll include the last document that breaches the threshold - if only `word_count_threshold` is specified. That was my line of reasoning.

Reply:
Okay, sure, makes sense. 👍  We just need to make sure that users can easily understand what the parameters do and for example what it means to set `truncate` to `False` and `word_count_threshold` to some value.

Reply:
I'm actually a little confused as to why we have these at all? 

In our other rankers we typically allow the user to restrict the number of documents that are returned with the `top_k` parameter. I think I would prefer to using that over adding two new parameters that feel more appropriate in a preprocessor. What do you think?

Reply:
Yeah, good question. To ensure control on the word count threshold because we can't do a word count threshold post-LITM, it needs to be built into the algorithm as it lays out documents in the LITM order.  You can't apply it post-LITM; it has to be during algorithm execution.

Reply:
Right, I see what you are saying. And I see how this is different compared to our other rankers. I think my preference would be to have the user control this through just the top_k alone for consistency with our other rankers, but I can see their usefulness so I'll leave it up to you for if you'd like to keep them.

One thing to point out though is that even with these parameters it does not guarantee that a future node like the PromptNode won't still further truncate the documents so this is still some possibility for error here.

Reply:
Ok, let's then introduce top_k and leave word_count_threshold as an option, but let's not truncate. Strikes a good balance, while not breaking existing paradigms.